src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -122,10 +122,16 @@ public static partial class JsonSerializer
                     ThrowHelper.ThrowJsonException_MetadataValueWasNotString(reader.TokenType);
                 }
 
-                string key = reader.GetString()!;
+                string referenceId = reader.GetString()!;
+                object value = state.ReferenceResolver.ResolveReference(referenceId);
+                Type typeOfValue = value.GetType();
 
-                // todo: https://github.com/dotnet/runtime/issues/32354
-                state.Current.ReturnValue = state.ReferenceResolver.ResolveReference(key);
+                if (!converter.TypeToConvert.IsAssignableFrom(value.GetType()))
+                {
+                    ThrowHelper.ThrowJsonException_MetadataReferenceOfTypeCannotBeAssignedToType(referenceId, typeOfValue, converter.TypeToConvert);
+                }
```

Comment:
This looks expensive. Does this affect perf? Would a try-catch-rethrow for `InvalidCastException` at the site of the assignment/setting the property (in a separate code path used when the reference handling feature is on) be more efficient?

Reply:
These are the perf results that involve this code path:

```
summary:
better: 2, geomean: 1.014
worse: 2, geomean: 1.020
total diff: 4

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| ReadPreservedReferences<IndexViewModel>.DeserializePreserved(IsDataPreserved: Tr |      1.02 |          8721.92 |          8932.26 |         |
| ReadPreservedReferences<MyEventsListerViewModel>.DeserializePreserved(IsDataPres |      1.02 |         30400.55 |         30896.48 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| ReadPreservedReferences<Location>.DeserializePreserved(IsDataPreserved: True)    |      1.02 |          1388.51 |          1362.30 |         |
| ReadPreservedReferences<LoginViewModel>.DeserializePreserved(IsDataPreserved: Tr |      1.01 |           686.08 |           680.12 |         |
```

Reply:
`IndexViewModel` and `MyEventsListerViewModel` are models with a heavy amount of duplicate references, hence they show a small regression.

Reply:
Using try-catch instead of the current approach does not help with performance either. 
```
summary:
worse: 4, geomean: 1.016
total diff: 4

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.ReadPreservedReferences<LoginViewModel>.Des |      1.02 |           662.57 |           678.43 |         |
| System.Text.Json.Serialization.Tests.ReadPreservedReferences<IndexViewModel>.Des |      1.02 |          8552.66 |          8698.37 |         |
| System.Text.Json.Serialization.Tests.ReadPreservedReferences<MyEventsListerViewM |      1.02 |         29335.54 |         29784.82 |         |
| System.Text.Json.Serialization.Tests.ReadPreservedReferences<Location>.Deseriali |      1.01 |          1351.19 |          1359.89 |         |

No Faster results for the provided threshold = 0% and noise filter = 0.3ns.
```

I also tried with generic `HandleMetadata<T>` and checking `!(value is T)` but that does not bring too much benefit as well.

```
summary:
better: 1, geomean: 1.008
worse: 2, geomean: 1.018
total diff: 3

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.ReadPreservedReferences<MyEventsListerViewM |      1.02 |         29739.36 |         30341.87 |         |
| System.Text.Json.Serialization.Tests.ReadPreservedReferences<Location>.Deseriali |      1.02 |          1349.55 |          1370.82 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.ReadPreservedReferences<IndexViewModel>.Des |      1.01 |          8752.77 |          8680.62 |         |
```

Reply:
Can we avoid calling `IsAssignableFrom()` when `IsInternalConverter==true`? That would at least prevent regression for standard fast-path scenarios.

Reply:
We use the slow path (that uses ObjectState) when the feature is on. Besides that, how does `IsInternalConverter` tell us if the `value` pulled from `ReferenceResolver` can be assigned to `converter.TypeToConvert`?